relay-event-normalization/src/normalize/span/tag_extraction.rs

```diff
@@ -354,11 +355,60 @@ static SQL_TABLE_EXTRACTOR_REGEX: Lazy<Regex> = Lazy::new(|| {
     Regex::new(r#"(?i)(from|into|update)(\s|")+(?P<table>(\w+(\.\w+)*))(\s|")+"#).unwrap()
 });
 
-/// Returns the table in the SQL query, if any.
-///
-/// If multiple tables exist, only the first one is returned.
-fn sql_table_from_query(query: &str) -> Option<&str> {
-    extract_captured_substring(query, &SQL_TABLE_EXTRACTOR_REGEX)
+/// Used to replace placeholders (parameters) by dummy values such that the query can be parsed.
+static SQL_PLACEHOLDER_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"(?:\?+|\$\d+|%(?:\(\w+\))?s|:\w+)").unwrap());
+
+/// Returns a sorted, comma-separated list of SQL tables, if any.
+fn sql_tables_from_query(db_system: Option<&str>, query: &str) -> Option<String> {
+    use sqlparser::ast::{ObjectName, Visit, Visitor};
+    use sqlparser::dialect::GenericDialect;
+    use std::ops::ControlFlow;
+    /// Visitor that finds relation names.
+    struct RelationsVisitor {
+        relations: BTreeSet<String>,
+    }
+
+    impl Visitor for RelationsVisitor {
+        type Break = ();
+
+        fn pre_visit_relation(&mut self, relation: &ObjectName) -> ControlFlow<Self::Break> {
+            if let Some(name) = relation.0.last() {
+                let last = name.value.split('.').last().unwrap_or(&name.value);
+                self.relations.insert(last.to_lowercase());
+            }
+            ControlFlow::Continue(())
+        }
+    }
```

Comment:
not that im against it, but im a bit curious on this pattern. I didn't see before having structs and impl blocks inside of a function, is this idiomatic ?

Reply:
The language allows it, and I like how it keeps the types private. But I'll move them out (see comment above).